torch/_export/verifier.py

```diff
@@ -420,6 +420,35 @@ def _verify_exported_program_signature(exported_program) -> None:
                 f"Unknown InputKind {input_spec.kind}."
             )
 
+    # Check the order of input spec kinds
+    expected_order = [
+        InputKind.TOKEN,
+        InputKind.PARAMETER,
+        InputKind.BUFFER,
+        InputKind.CONSTANT_TENSOR,
+        InputKind.CUSTOM_OBJ,
+        InputKind.USER_INPUT,
+    ]
+    current_order_index = 0
+    seen_non_persistent_buffer = False
+    for input_spec in gs.input_specs:
+        kind_index = expected_order.index(input_spec.kind)
+        if kind_index < current_order_index:
+            raise SpecViolationError(
+                f"InputSpec kind {input_spec.kind} is out of order. Expected order is {expected_order}."
```

Comment:
I believe this is BC breaking if you add the verifier check for all exported programs, because saved exported programs who do not have this ordering will fail that check when loading.